usr/libexec/helper-scripts/accountctl.sh

```diff
@@ -0,0 +1,352 @@
+## Example: get_clean_pass user '!*'
+get_clean_pass(){
+  log info "${FUNCNAME[0]} $*"
+  local user pass symbol
+  user="${1:-}"
+  is_user "${user}"
+  symbol="${2:-"!*"}"
+  pass="$(get_pass "${user}")"
+  pass="${pass#"${symbol}"}"
+  printf '%s\n' "${pass}"
+}
+
+
+## Description: Check if password is empty.
+## Output: None
+## Return: 0 when empty, 1 otherwise.
+## Usage: is_pass_empty USER
+## Example: is_pass_empty user
+is_pass_empty(){
+  log info "${FUNCNAME[0]} $*"
+  local user trim_pass
+  user="${1:-}"
+  is_user "${user}"
+  trim_pass="$(get_clean_pass "${user}" '!*')"
+  if test -z "${trim_pass}"; then
+    return 0
+  fi
+  return 1
+}
+
+
+## Description: Check if password is locked.
+## Output: None
+## Return: 0 when locked, 1 otherwise.
+## Usage: is_pass_locked USER
+## Example: is_pass_locked user
+is_pass_locked(){
+  log info "${FUNCNAME[0]} $*"
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  case "$(get_pass "${user}")" in
+    "!"*) return 0;;
+    *) return 1;;
+  esac
+}
+
+
+## Description: Check if password is disabled.
+## Output: None
+## Return: 0 when disabled, 1 otherwise.
+## Usage: is_pass_disabled USER
+## Example: is_pass_disabled user
+is_pass_disabled(){
+  log info "${FUNCNAME[0]} $*"
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  case "$(get_pass "${user}")" in
+    "*"*|"!*"*) return 0;;
+    *) return 1;;
+  esac
+}
+
+
+## Description: Lock a password.
+## Output: None
+## Return: passwd
+## Usage: lock_pass USER
+## Example: lock_pass user
+lock_pass(){
+  log info "${FUNCNAME[0]} $*"
+  has passwd
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  if is_pass_locked "${user}"; then
+    return 0
+  fi
+  passwd -ql -- "${user}"
+}
+
+
+## Description: Unlock a password.
+## Output: None
+## Return: chpasswd
+## Usage: unlock_pass USER
+## Example: unlock_pass user
+unlock_pass(){
+  log info "${FUNCNAME[0]} $*"
+  has chpasswd
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  if ! is_pass_locked "${user}"; then
+    return 0
+  fi
+  ## The tools "passwd" and "usermod" can't unlock when password is empty.
+  pass="$(get_clean_pass "${user}" '!')"
+  chpasswd -c NONE <<< "${user}:${pass}"
+}
+
+
+## Description: Disable a password.
+## Output: None
+## Return: chpasswd
+## Usage: disable_pass USER
+## Example: disable_pass user
+disable_pass(){
+  log info "${FUNCNAME[0]} $*"
+  has chpasswd
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  if is_pass_disabled "${user}"; then
+    return 0
+  fi
+  pass="$(get_pass "${user}")"
+  chpasswd -c NONE <<< "${user}:*${pass}"
+}
+
+
+## Description: Enable a password.
+## Output: None
+## Return: chpasswd
+## Usage: enable_pass USER
+## Example: enable_pass user
+enable_pass(){
+  log info "${FUNCNAME[0]} $*"
+  has chpasswd
+  local user
+  user="${1:-}"
+  is_user "${user}"
+  if ! is_pass_disabled "${user}"; then
+    return 0
+  fi
+  pass="$(get_clean_pass "${user}" '*')"
+  chpasswd -c NONE <<< "${user}:${pass}"
```

Comment:
This will fail to remove the `*` in the locked-and-disabled (`!*`) edge case. One has to use `unlock_pass` first and then `enable_pass` for things to work, if they use this in the reverse order the password will be left unlocked but still disabled.

One solution here might be to check with `is_pass_locked` too, and in that edge case filter out the entirety of the `!*` sequence, then add the `!` back to avoid unlocking the password incorrectly:

```suggestion
  pass="$(get_clean_pass "${user}" '*')"
  if is_pass_locked "${user}"; then
    pass="$(get_clean_pass "${user}" '!*')"
    chpasswd -c NONE <<< "${user}:!${pass}"
  else
    pass="$(get_clean_pass "${user}" '*')"
    chpasswd -c NONE <<< "${user}:${pass}"
  fi
```